syntax = "proto3";

message TrialTransactionPayload {

    enum PayloadType {
        TYPE_UNSET = 0;
        CREATE_SPONSOR = 1;
        CREATE_INVESTIGATOR = 2;
        CREATE_PATIENT = 3;
        CREATE_HOSPITAL = 4;
        CREATE_EHR = 5;
//        CREATE_DATA_PROVIDER = 6;
        IMPORT_DATA = 6;
        UPDATE_DATA = 7;
        SET_ELIGIBLE = 8;
//        NEXT_VISIT = 10;
//        ADD_LAB_TEST = 11;
//        ADD_PULSE = 12;
//        CREATE_LAB = 13;
//        UPDATE_CLAIM = 14;
//        CLOSE_CLAIM = 15;
    }

    PayloadType payload_type = 1;

    Patient create_patient = 2;
    Hospital create_hospital = 3;
    EHR create_ehr = 4;
//    DataProvider create_data_provider = 5;
    Sponsor create_sponsor = 6;
    Investigator create_investigator = 7;
    Data import_data = 8;
    Data update_data = 9;
    Data set_eligible = 10;
//    Claim create_claim = 5;
//    ActionOnClaim assign_doctor = 6;
//    ActionOnClaim first_visit = 7;
//    ActionOnClaim pass_tests = 8;
//    ActionOnClaim attend_procedures = 9;
//    ActionOnClaim eat_pills = 10;
//    ActionOnClaim next_visit = 11;
//    AddLabTest lab_test = 12;
//    AddPulse pulse = 13;
//    CreateLab create_lab = 14;
//    Claim close_claim = 15;
//    Claim update_claim = 16;
}

message Patient {
    string public_key = 1;
    string name = 2;
    string surname = 3;
}

message Hospital {
    string public_key = 1;
    string name = 2;
}

message EHR {
    string id=1;
    string client_pkey=2;
    string height=3;
    string weight=4;
    string A1C=5;
    string FPG=6;
    string OGTT=7;
    string RPGT=8;
    string event_time=9;
}

message EHRWithUser {
    string id=1;
    string client_pkey=2;
    string height=3;
    string weight=4;
    string A1C=5;
    string FPG=6;
    string OGTT=7;
    string RPGT=8;
    string event_time=9;
    string name=10;
    string surname=11;
}

//message DataProvider {
//    string public_key = 1;
//    string name = 2;
//}

message Data {
    string id=1;
    string height=2;
    string weight=3;
    string A1C=4;
    string FPG=5;
    string OGTT=6;
    string RPGT=7;
    bool eligible=8;
    string event_time=9;
}

message Data {
    string id=1;
    string field_1=3;
    string field_2=4;
    string event_time=5;
}

message Sponsor { //  Study management

    enum Type {
        CRO = 0;
        Pharma = 1;
        Medtech = 2;
    }

    string public_key = 1;
    string name = 2;
}

message Investigator {
    string public_key = 1;
    string name = 2;
    string surname = 3;
}



//message CreateLab {
//    string public_key = 1;
//    string name = 2;
////    repeated Permission permissions = 2;
//}
//
//message AddLabTest {
//    string height=1;
//    string weight=2;
//    string gender=3;
//    string a_g_ratio=4;
//    string albumin=5;
//    string alkaline_phosphatase=6;
//    string appearance=7;
//    string bilirubin=8;
//    string casts=9;
//    string color=10;
//    string event_time=11;
//    string id=12;
//    string client_pkey=13;
//}
//
//message AddLabTestWithUser {
//    string height=1;
//    string weight=2;
//    string gender=3;
//    string a_g_ratio=4;
//    string albumin=5;
//    string alkaline_phosphatase=6;
//    string appearance=7;
//    string bilirubin=8;
//    string casts=9;
//    string color=10;
//    string event_time=11;
//    string id=12;
//    string client_pkey=13;
//    string name=14;
//    string surname=15;
//}
//
//message AddPulse {
////    string public_key=1;
//    string id=1;
//    string pulse=2;
//    string timestamp=3;
//    string client_pkey=4;
//}
//
//message AddPulseWithUser {
////    string public_key=1;
//    string id=1;
//    string pulse=2;
//    string timestamp=3;
//    string client_pkey=4;
//    string name=5;
//    string surname=6;
//}
//
//message CreateClaim {
//    string claim_id = 1;
//    string clinic_pkey = 2;
//    string patient_pkey = 3;
//}
//
//message Claim {
//
//    enum State {
//        OPENED = 0;
//        CLOSED = 1;
//    }
//
//    string id = 1;
//    string client_pkey = 2;
//    string description = 3;
//    string provided_service = 4;
//    State state = 5;
//    string contract_id = 6;
//}
//
//message ClaimWithUser {
//
//    enum State {
//        OPENED = 0;
//        CLOSED = 1;
//    }
//
//    string id = 1;
//    string client_pkey = 2;
//    string description = 3;
//    string provided_service = 4;
//    State state = 5;
//    string contract_id = 6;
//    string name = 7;
//    string surname = 8;
//}
//
//message ActionOnClaim {
//    enum Event {
//        ASSIGN = 0;
//        FIRST_VISIT = 1;
//        PASS_TEST = 2;
//        PASS_PROCEDURE = 3;
//        EAT_PILLS = 4;
//        NEXT_VISIT = 5;
//    }
//
//    string claim_id = 1;
//    string clinic_pkey = 2;
//    string description = 3;
//    Event event = 4;
//    string event_time = 5;
//}